src/Microsoft.Health.Fhir.SqlServer/Features/Search/Expressions/Visitors/IncludeRewriter.cs

```diff
@@ -55,12 +50,116 @@ public override Expression VisitSqlRoot(SqlRootExpression expression, object con
                 }
             }
 
-            if (containsInclude)
+            reorderedExpressions.Add(IncludeUnionAllExpression);
+            return new SqlRootExpression(reorderedExpressions, expression.DenormalizedExpressions);
+        }
+
+        private class TableExpressionComparer : IComparer<TableExpression>
+        {
+            public int Compare(TableExpression x, TableExpression y)
```

Comment:
Yes, the purpose was to achieve topological sort of dependencies. 
What indirect dependency do you have in mind? I couldn't think of any.
Do you have an example of a query with indirect dependency?

Can we accept it in this PR and open an issue to implement a topological sort?

Reply:
Yes, I was trying to achieve topological sort. I see your point, but I don't see how we miss an indirect dependency with iterate. 
Say we have a < b < c search parameters. 

> In `cte4` below, we apply a top clause on the resultset from `cte3` we use for the intersection:
> 
> ```sql
>       cte3 AS
>       (
>           SELECT DISTINCT Sid1, IsMatch, 0 AS IsPartial
>           FROM cte2
>       ),
>       cte4 AS
>       (
>           SELECT DISTINCT refTarget.ResourceSurrogateId AS Sid1, 0 AS IsMatch
>           FROM dbo.ReferenceSearchParam refSource
>           INNER JOIN dbo.Resource refTarget
>           ON refSource.ReferenceResourceTypeId = refTarget.ResourceTypeId
>               AND refSource.ReferenceResourceId = refTarget.ResourceId
>           WHERE refSource.SearchParamId = @p5
>               AND refTarget.IsHistory = 0
>               AND refSource.IsHistory = 0
>               AND refSource.ResourceTypeId = @p6
>               AND refSource.ResourceSurrogateId IN (SELECT TOP(@p7) Sid1 FROM cte3)
>       )
> ```
> 
> We should only be applying iterate to the first ten includes in `cte3`. (We've also discussed that we should apply an overall limit to included resources like we do for `_revinclude`, though I'm happy to take that as a separate PR)

True, but we agreed to have a separate issue for it: https://github.com/microsoft/fhir-server/issues/1309

Reply:
Regarding the TOP clause, I'm sorry my comment forgot a "not". We should not be applying the top clause here. We do not want to limit this to the first 10 includes.

Reply:
Right! Now it makes sense. I think we should limit it to the IncludeLimit though (default 100)

Reply:
As for topological sort, this will not work when we introduce a recursive queries (circular reference), but the comparator could actually work since we always put :iterate after _include/_revinclude. 

For example, this is a recursive query with circular reference since participant can be of type Practitioner. 
The comparator will actually help. 
CareTeam resource?
CareTeam?_include=CareTeam:participant&_include:iterate=Patient:general-practitioner

I'll try to see if I can create a test that fails on indirect dependency.
So for 3 parameters where a < b < c - I basically need to check all the possible queries (order permutations).
If all pass than I think we are safe with comparer

Reply:
We should maybe limit it, but then I suppose we should make sure we raise the "is partial" warning. Or we put no limit and let #1309 take care of limiting it.

Reply:
Looks like we are safe (for now).
Tried all 6 permutations of 3 parameters in this query and it was sorted correctly:
MedicationDispense?_include=MedicationDispense:prescription&_include:iterate=MedicationRequest:patient&_include:iterate=Patient:general-practitioner

Reply:
I think this is an example where this approach fails.
Correct results:
```
https://localhost:44348/MedicationDispense?_id=1&_include:iterate=Device:location&_include:iterate=Location:endpoint&_include=MedicationDispense:performer&_include:iterate=Patient:general-practitioner
```
Incorrect results (I just moved `_include:iterate=Device:location` to the back):
```
https://localhost:44348/MedicationDispense?_id=1&_include:iterate=Location:endpoint&_include=MedicationDispense:performer&_include:iterate=Patient:general-practitioner&_include:iterate=Device:location
```
[correct.txt](https://github.com/microsoft/fhir-server/files/5374716/correct.txt)
[incorrect.txt](https://github.com/microsoft/fhir-server/files/5374717/incorrect.txt)

Reply:
(BTW: I think it should be an error when you :iterate over a type that is not part the rest of the query)

Reply:
I think that's not a 'sort' issue, but as you said, it's due to having an "orphan" parameters that are not related to the query/

I investigated this case  when I started working on the feature, and this is how I found this issue: https://github.com/microsoft/fhir-server/issues/1267 
I left it out of scope since it is not an :iterate issue necessarily, but an  overall _include/_revinclude issue.

Hapi for example, doesn't throw an error, but rather ignores the "orphan" parts.

If we want to through an error, we could actually check for such "orphans" in the IncludeRewriter.

I don't think this should be handled in this specific PR.

Thoughts?

Reply:
Agree.

Reply:
Topological sort was implemented to order the parameters. 
Also, I have added the above queries to the unit tests. Note that all the parameters above are actually related to the query. Turns out MedicationDispense:performer includes Device.